packages/create-turbo/src/commands/create/createProject.ts

```diff
@@ -0,0 +1,231 @@
+import retry from "async-retry";
+import chalk from "chalk";
+import fs from "fs-extra";
+import path from "path";
+import semverPrerelease from "semver/functions/prerelease";
+
+import {
+  downloadAndExtractExample,
+  downloadAndExtractRepo,
+  getRepoInfo,
+  existsInRepo,
+  hasRepo,
+  RepoInfo,
+} from "../../utils/examples";
+import { addGitIgnore } from "../../utils/git";
+import { isFolderEmpty } from "../../utils/isFolderEmpty";
+import { isWriteable } from "../../utils/isWriteable";
+import { turboLoader, error } from "../../logger";
+import cliPkgJson from "../../../package.json";
+
+export class DownloadError extends Error {}
+
+export async function createProject({
+  appPath,
+  projectName,
+  example,
+  examplePath,
+}: {
+  appPath: string;
+  projectName: string;
+  example: string;
+  examplePath?: string;
+}): Promise<{
+  cdPath: string;
+  hasPackageJson: boolean;
+  availableScripts: Array<string>;
+}> {
+  let repoInfo: RepoInfo | undefined;
+  let repoUrl: URL | undefined;
+  const isDefaultExample = example === "basic" || example === "default";
+
+  try {
+    repoUrl = new URL(example);
+  } catch (err: any) {
+    if (err.code !== "ERR_INVALID_URL") {
+      error(err);
+      process.exit(1);
+    }
+  }
+
+  if (repoUrl) {
+    if (repoUrl.origin !== "https://github.com") {
+      error(
+        `Invalid URL: ${chalk.red(
+          `"${example}"`
+        )}. Only GitHub repositories are supported. Please use a GitHub URL and try again.`
+      );
+      process.exit(1);
+    }
+
+    repoInfo = await getRepoInfo(repoUrl, examplePath);
+
+    if (!repoInfo) {
+      error(
+        `Found invalid GitHub URL: ${chalk.red(
+          `"${example}"`
+        )}. Please fix the URL and try again.`
+      );
+      process.exit(1);
+    }
+
+    const found = await hasRepo(repoInfo);
+
+    if (!found) {
+      error(
+        `Could not locate the repository for ${chalk.red(
+          `"${example}"`
+        )}. Please check that the repository exists and try again.`
+      );
+      process.exit(1);
+    }
+  } else {
+    const found = await existsInRepo(example);
+
+    if (!found) {
+      error(
+        `Could not locate an example named ${chalk.red(
+          `"${example}"`
+        )}. It could be due to the following:\n`,
+        `1. Your spelling of example ${chalk.red(
+          `"${example}"`
+        )} might be incorrect.\n`,
+        `2. You might not be connected to the internet or you are behind a proxy.`
+      );
```

Comment:
Two very different types of errors. Seems like something we could throw/catch different known errors instead of this? (maybe i'm getting too used to Golang 😃 )

Reply:
This is straight from Next, I'll look into improving it